### PR TITLE
Improve Extension Authoring Experience with new Build Tasks and Launch Configs

### PR DESCRIPTION
**Pull request by JustinGrote**
## PR Summary

Updates the build tasks and launch configs to:
- Massively shorten inner loop development on typescript features by using incremental watch configs
- Integrate linting process into vscode so that lint errors are surfaced in the IDE prior to CI commit or manual lint run
- Enable semicolons for typescript actions like auto-formatting since the current ESLint profile requires them.
- Provide a launch test task for isolated testing separate from developer's profile
- Provide extension launchers for interactive debugging of the extension in either the developer's profile, a temp throwaway profile, or an isolated but persistent profile
image

## Testing

1. Open the workspace file
2. Run ./build at least once (integrated PSES watch builds and auto-attach debugging is future PR)
3. Click the `Test Extension` Launch Task and note the vscode instance that starts does not use your personal settings
4. Click the `Launch Extension` Task and note the launched extension has your settings. First run may take a while but future starts should be near-instant. Also note that there are watch tasks now running for both tests files and build files and will auto-update every time you save a change and have a status indicator to the right.
image

6. Click the `Launch Extension - Isolated Profile` task and note the vscode window is isolated. Change your theme and launch again, observe that theme persisted but did not change developer's personal settings.
7. Repeat with `Launch Extension - Temp` but note the theme change does not persist.
8. Insert an intentional error into a .ts file and try to run a launch config. An error will appear saying that errors ocurred and will take you straight to the errors.
image

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

File: extension-dev.code-workspace [hunk 1]
```diff
@@ -21,6 +21,8 @@
     ]
   },
   "settings": {
+    "window.title": "PowerShell VSCode Extension Development",
```
Comment (andschwa) [inline extension-dev.code-workspace:24]:
```suggestion
    "window.title": "PowerShell VS Code Extension Development",
```

File: extension-dev.code-workspace [hunk 2]
```diff
@@ -44,7 +46,17 @@
     "powershell.codeFormatting.whitespaceBetweenParameters": true,
     "powershell.codeFormatting.pipelineIndentationStyle": "IncreaseIndentationForFirstPipeline",
     // Lock the TypeScript SDK path to the version we use
-    "typescript.tsdk": "Client/node_modules/typescript/lib"
+    "typescript.tsdk": "Client/node_modules/typescript/lib",
+    // Code actions like "organize imports" ignore eslint, so we need this here
```
Comment (andschwa) [inline extension-dev.code-workspace:50]:
```suggestion
    // Code actions like "organize imports" ignore ESLint, so we need this here
```

File: extension-dev.code-workspace [hunk 3]
```diff
@@ -90,7 +102,7 @@
         "options": {
           "cwd": "${workspaceFolder:Client}"
         },
-        "command": "Invoke-Build Build",
+        "command": "./build.ps1",
```
Comment (andschwa) [inline extension-dev.code-workspace:105]: What's the reason for this change?

File: extension-dev.code-workspace [hunk 4]
```diff
@@ -197,42 +286,129 @@
...
+        // Runs the extension in an empty temp profile that is automatically cleaned up after use
+        // Undocumented: https://github.com/microsoft/vscode-docs/issues/6220
+        "name": "Launch Extension - Temp Profile",
         "type": "extensionHost",
         "request": "launch",
         "runtimeExecutable": "${execPath}",
         "args": [
-          "--disable-extensions",
-          "--extensionDevelopmentPath=${workspaceFolder:Client}"
+          "--profile-temp",
```
Comment (andschwa) [inline extension-dev.code-workspace:304]: Cool!

File: extension-dev.code-workspace [hunk 5]
```diff
@@ -44,7 +46,17 @@
...
+    // Code actions like "organize imports" ignore eslint, so we need this here
+    "typescript.format.semicolons": "insert",
+    // Enable ESLint as defaut formatter so quick fixes can be applied directly
+    "eslint.format.enable": true,
+    "[typescript]": {
+      "editor.defaultFormatter": "dbaeumer.vscode-eslint",
+      "editor.formatOnPaste": true,
+      "editor.formatOnSave": true,
+      "editor.formatOnSaveMode": "modificationsIfAvailable"
+    }
```
Comment (andschwa) [inline extension-dev.code-workspace:59]: Love it.

File: extension-dev.code-workspace [hunk 6]
```diff
@@ -184,7 +228,52 @@
...
+        "request": "launch",
+        "runtimeExecutable": "${execPath}",
+        "args": [
+          "--extensionDevelopmentPath=${workspaceFolder:Client}"
+        ],
+        "env": {
+          "__TEST_WORKSPACE_PATH": "${workspaceFolder:Client}/examples",
+        },
+        "sourceMaps": true,
+        // This speeds up source map detection and makes smartStep work correctly
```
Comment (andschwa) [inline extension-dev.code-workspace:260]: Oh thank you.

File: extension-dev.code-workspace [hunk 7]
```diff
@@ -197,42 +286,129 @@
...
+          "hidden": false,
+          "group": "test",
+          "order": 2
+        }
       },
       {
-        "name": "Launch Extension Tests",
+        // Runs the extension in an isolated but persistent profile separate from the user settings
+        // Undocumented: https://github.com/microsoft/vscode-docs/issues/6220
+        "name": "Launch Extension - Isolated Profile",
```
Comment (andschwa) [inline extension-dev.code-workspace:329]: What's the use case of this versus the temporary profile?

File: package.json [hunk 8]
```diff
@@ -115,8 +115,11 @@
   "main": "./out/main.js",
   "scripts": {
     "lint": "eslint . --ext .ts",
-    "build": "tsc --project tsconfig.json && esbuild ./src/main.ts --outdir=out --bundle --external:vscode --platform=node",
-    "test": "node ./out/test/runTests.js",
+    "build": "esbuild ./src/main.ts --outdir=out --bundle --external:vscode --platform=node",
+    "build-watch": "npm run build -- --watch",
+    "build-test": "tsc --incremental",
+    "build-test-watch": "npm run build-test -- --watch",
+    "test": "npm run build-test && node ./out/test/runTests.js",
```
Comment (andschwa) [inline package.json:122]: Cool!

File: test/.vscode/settings.json
Comment (andschwa) [inline test/.vscode/settings.json:1]: Hurray!

File: test/TestEnvironment.code-workspace [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
+  // A simple test environment that suppresses some first start warnings we dont care about.
```
Comment (andschwa) [inline test/TestEnvironment.code-workspace:2]:
```suggestion
  // A simple test environment that suppresses some first start warnings we don't care about.
```

File: vscode-powershell.build.ps1 [hunk 10]
```diff
@@ -113,7 +113,7 @@ task Test Build, {
     Write-Host "`n### Running extension tests" -ForegroundColor Green
     Invoke-BuildExec { & npm run test }
     # Reset the state of files modified by tests
-    Invoke-BuildExec { git checkout package.json test/.vscode/settings.json}
```
Comment (andschwa) [inline vscode-powershell.build.ps1:116]: Is this one getting modified by tests too now?

**Review by andschwa — CHANGES_REQUESTED** [review]
I like it!

File: extension-dev.code-workspace (hunk 3 above)
Comment (JustinGrote) [reply extension-dev.code-workspace:105]: build.ps1 has more scaffolding currently than Invoke-Build does. If we plan to make the invoke-build task all-encompassing and just make build.ps1 an entrypoint then this can change back, I plan to evaluate that stuff in a future PR.

File: extension-dev.code-workspace (hunk 5 above)
Comment (JustinGrote) [reply extension-dev.code-workspace:59]: There will be more on the way, I plan to add some more eslint plugins for mocha (like catching the fact I keep forgetting to remove my .only, etc.), it's a good start though :)

File: extension-dev.code-workspace (hunk 7 above)
Comment (JustinGrote) [reply extension-dev.code-workspace:329]: It's persistent. Temp profile is thrown away each time, in this one you can test some persistent stuff interactively, like lets say you want to manually set up some settings to test something, it'll persist and then you can just delete it after.

File: vscode-powershell.build.ps1 (hunk 10 above)
Comment (JustinGrote) [reply vscode-powershell.build.ps1:116]: I don't think the package.json was getting modified before, it might be cruft, but I didn't want to remove it just in case. It doesn't hurt anything to have it here.

File: extension-dev.code-workspace (hunk 3 above)
Comment (andschwa) [reply extension-dev.code-workspace:105]: Sounds good!

File: extension-dev.code-workspace (hunk 7 above)
Comment (andschwa) [reply extension-dev.code-workspace:329]: Ah I see!

File: vscode-powershell.build.ps1 (hunk 10 above)
Comment (andschwa) [reply vscode-powershell.build.ps1:116]: There was a while that the VS Code test package was modifying that file and it seemed to be messing stuff up somehow, hence the origin of this. It can probably go away but yeah also it doesn't hurt anything.

**Review by andschwa — APPROVED** [review]
